File: src/queue/ts/dynamicQueue.ts
import { INode } from "./types"
import { EmptyQueueError } from "./utils"

class Node<Type> implements INode<Type> {
    constructor(
        public value: Type,
        public next: INode<Type> | null
    ) {}
}

export class DynamicQueue<Type> {
    private head: INode<Type> | null
    private tail: INode<Type> | null

    constructor() {
        this.head = null
        this.tail = null
    }

    public enqueue(value: Type): void {
        const node = new Node(value, null)

        if (this.head === null || this.tail === null) {
            this.head = node
            this.tail = node
        } else {
            this.tail.next = node
            this.tail = node
        }
    }

    public dequeue(): Type {
        if (this.head === null || this.tail === null) {
            throw new EmptyQueueError()
        }

        const value = this.head.value
        this.head = this.head.next
        return value
    }

    public peek(): Type {
        if (this.head === null) {
            throw new EmptyQueueError()
        }
        return this.head?.value
    }

    public print(): void {
        if (this.isEmpty()) {
            throw new EmptyQueueError()
        }

        let current = this.head
        let output = ""

        while (current !== null) {
            output += `(${current.value})-`
            current = current.next
        }

        console.log(output.slice(0, -1))
    }

    public isEmpty(): boolean {
        return this.head === null
    }
}

File: src/queue/c/dynamicQueue.c
#include <stdio.h>
#include <stdlib.h>
#include "dynamicQueue.h"


struct DynamicQueue* createQueue(){
    struct DynamicQueue *queue = malloc(sizeof(struct DynamicQueue));
    queue->head = NULL;
    queue->tail = NULL;

    return queue;
}

int enqueue(struct DynamicQueue *queue, int value){
    struct Node *current = queue->head;
    struct Node *node = malloc(sizeof(struct Node));

    node->value = value;
    node->next = NULL;

    if (isEmpty(queue)){
        queue->head = node;
        queue->tail = node;
    } else {
        queue->tail->next = node;
        queue->tail = node;
    }

    return 0;
}

int dequeue(struct DynamicQueue *queue){
    if (isEmpty(queue)){
        printf("Queue is empty!\n");
        return -1;
    }

    struct Node *node = queue->head;
    int value = node->value;
    queue->head = queue->head->next;

    free(node);

    return value;
}

int peek(struct DynamicQueue *queue){
    if (isEmpty(queue)){
        printf("Queue is empty!\n");
        return -1;
    }

    return queue->head->value;
}

int print(struct DynamicQueue *queue){
    if (isEmpty(queue)){
        printf("Queue is empty!\n");
        return -1;
    }

    struct Node *current = queue->head;

    while (current != NULL){
        printf("(%d)", current->value);
        if (current->next != NULL){
            printf("-");
        }
        current = current->next;
    }
    printf("\n\n");

    return 0;
}

int isEmpty(struct DynamicQueue *queue){
    return queue->head == NULL;
}
